### PR TITLE
Gizmo line styles

### PR DESCRIPTION
**Pull request by solis-lumine-vorago**
# Objective

- Adds line styles to bevy gizmos, suggestion of #9400 
- Currently solid and dotted lines are implemented but this can easily be extended to support dashed lines as well if that is wanted.

## Solution

- Adds the enum `GizmoLineStyle` and uses it in each `GizmoConfig` to configure the style of the line.
- Each "dot" in a dotted line has the same width and height as the `line_width` of the corresponding line.

---

## Changelog

- Added `GizmoLineStyle` to `bevy_gizmos`
- Added a `line_style: GizmoLineStyle ` attribute to `GizmoConfig`
- Updated the `lines.wgsl` shader and the pipelines accordingly.

## Migration Guide

- Any manually created `GizmoConfig` must now include the `line_style` attribute

## Additional information
Some pretty pictures :)

This is the 3d_gizmos example with/without `line_perspective`:
Screenshot 2024-03-09 at 23 25 53
Screenshot 2024-03-09 at 23 25 39

And the 2d example:
Screenshot 2024-03-09 at 23 25 06

### CODE REVIEW

**Comment by pablo-lua** [comment]
At this point I think we should break gizmos 2d_example and 3d_example in a few parts, the examples are getting pretty cramped

**Comment by solis-lumine-vorago** [comment]
> At this point I think we should break gizmos 2d_example and 3d_example in a few parts, the examples are getting pretty cramped

I agree but I think this should wait until both #12252 and this are merged so we can have one "fancy_gizmos" example.

File: crates/bevy_gizmos/src/config.rs
Comment (BD103) [inline crates/bevy_gizmos/src/config.rs:18]:
```suggestion
#[derive(Copy, Clone, Debug, Default, Hash, PartialEq, Eq, Reflect)]
#[non_exhaustive]
pub enum GizmoLineStyle {
```

It might be good to add this, so it's not a breaking change in the future if we add more line styles.

Comment (pablo-lua) [reply crates/bevy_gizmos/src/config.rs:18]: I think non_exhaustive can be more surprising than just a migration guide explaining what line style was added, or what was changed.

Comment (solis-lumine-vorago) [reply crates/bevy_gizmos/src/config.rs:18]: > I think non_exhaustive can be more surprising than just a migration guide explaining what line style was added, or what was changed.

I agree, especially since there is no real use case for matching against this enum. You are most likely just going to set it once or twice and not do anything else with it.

**Review by pablo-lua — APPROVED** [review]
All good. The problem with examples being too big can be in another PR.

File: crates/bevy_gizmos/src/lines.wgsl
Comment (BD103) [inline crates/bevy_gizmos/src/lines.wgsl:53]:
```suggestion

```

There appears to be some leftover whitespace. :)

File: crates/bevy_gizmos/src/lines.wgsl
Comment (BD103) [inline crates/bevy_gizmos/src/lines.wgsl:102]:
```suggestion
```

This extra newline can probably be deleted as well.

**Review by BD103 — APPROVED** [review]
Looks good and tested, though take my review with a grain of salt. I have little experience with gizmos or shaders. 😅

**Comment by alice-i-cecile** [comment]
Thank you to everyone involved with the authoring or reviewing of this PR! This work is relatively important and needs release notes! Head over to https://github.com/bevyengine/bevy-website/issues/1290 if you'd like to help out.

**Pull request merged**
